tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,146 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Discriminative Layer Training Optimizer for TensorFlow."""
+
+import tensorflow as tf
+from typeguard import typechecked
+from typing import Union
+from tensorflow.keras.optimizers import Optimizer
+
+# python -m flake8 tensorflow_addons/optimizers/discriminative_layer_training.py
+# python -m black tensorflow_addons/optimizers/discriminative_layer_training.py
+
+
+class FakeVar:
+    def __init__(self, name):
+        # probably can be refactored out
+        self.name = name
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class MultiOpt(Optimizer):
+    @typechecked
+    def __init__(
+        self,
+        optimizer_layer_pairs: Union[list, None] = None,
+        optimizer_specs: Union[list, None] = None,
+        name: str = "MultiOpt",
+        **kwargs
+    ):
+
+        """
+        Creates a wrapper around a set of instantiated optimizer layer pairs.
+
+        Each optimizer will optimize only the weights associated with its paired layer. This can be used
+        to implement discriminative layer training by assigning different learning rates to each optimizer
+        layer pair. (Optimizer, list(Layers)) pairs are also supported.
+
+        Currently, MultiOpt does not support callbacks that modify optimizers. However, you can instantiate
+        optimizer layer pairs with tf.keras.optimizers.schedules.LearningRateSchedule instead of a static learning
+        rate.
+
+        This code should function on CPU, GPU, and TPU.
+
+        Example:
+
+        ```python
+
+        model = get_model()
+
+        opt1 = tf.keras.optimizers.Adam(learning_rate=1e-4)
+        opt2 = tf.keras.optimizers.Adam(learning_rate=1e-2)
+
+        opt_layer_pairs = [(opt1, model.layers[0]),
+                            (opt2, model.layers[1:])]
+
+        loss = tf.keras.losses.MSE
+        optimizer = MultiOpt(opt_layer_pairs)
+
+        model.compile(optimizer=optimizer, loss = loss)
+
+        model.fit(x,y)
+
+        ```
+
+
+        """
+
+        super(MultiOpt, self).__init__(name, **kwargs)
+
+        if optimizer_specs is None and optimizer_layer_pairs is not None:
+            self.optimizer_specs = [
+                self.create_optimizer_spec(opt, layer)
+                for opt, layer in optimizer_layer_pairs
+            ]
+
+        elif optimizer_specs is not None and optimizer_layer_pairs is None:
+            self.optimizer_specs = optimizer_specs
+
+        else:
+            raise RuntimeError(
+                "You must specify either an list of optimizer_layer_pairs or a list of optimizer_specs"
+            )
+
+        self.initialized_optimizer_specs = [
+            self.initialize_from_optimizer_spec(spec) for spec in self.optimizer_specs
```

Comment:
to prevent serialization issues when running on TPU. I don't want to pass instantiated optimizers for graph construction on the TPU. This should protect from weird problems on serialization and deserialization. As in, when we need to initialize the multiopt from a serialized state, we cannot pass instantiated optimizers to the multiopt

Reply:
Thanks for the clarification 😃 It makes sense to me! Can we also have this in docstring? Thank you!

Reply:
so I rewrote some stuff that line is no longer needed after fixing the serialization